pkg/k8s/statedb.go

```diff
@@ -0,0 +1,329 @@
+	DefaultBufferWaitTime = 50 * time.Millisecond
+)
+
+func (cfg ReflectorConfig[Obj]) getBufferSize() int {
+	if cfg.BufferSize == 0 {
+		return DefaultBufferSize
+	}
+	return cfg.BufferSize
+}
+
+func (cfg ReflectorConfig[Obj]) getWaitTime() time.Duration {
+	if cfg.BufferWaitTime == 0 {
+		return DefaultBufferWaitTime
+	}
+	return cfg.BufferWaitTime
+}
+
+type k8sReflector[Obj any] struct {
+	ReflectorConfig[Obj]
+
+	initDone func(statedb.WriteTxn)
+	db       *statedb.DB
+	table    statedb.RWTable[Obj]
+}
+
+func (r *k8sReflector[Obj]) run(ctx context.Context, health cell.Health) error {
+	type entry struct {
+		deleted   bool
+		name      string
+		namespace string
+		obj       Obj
+	}
+	type buffer struct {
+		replaceItems []any
+		entries      map[string]entry
+	}
+	bufferSize := r.getBufferSize()
+	waitTime := r.getWaitTime()
+	table := r.table
+
+	transform := r.Transform
+	if transform == nil {
+		// No provided transform function, use the identity function instead.
+		transform = TransformFunc[Obj](func(obj any) (Obj, bool) { return obj.(Obj), true })
+	}
+
+	queryAll := r.QueryAll
+	if queryAll == nil {
+		// No query function provided, use All()
+		queryAll = QueryAllFunc[Obj](func(txn statedb.ReadTxn, tbl statedb.Table[Obj]) statedb.Iterator[Obj] {
+			return tbl.All(txn)
+		})
+	}
+
+	// Construct a stream of K8s objects, buffered into chunks every [waitTime] period
+	// and then committed.
+	// This reduces the number of write transactions required and thus the number of times
+	// readers get woken up, which results in much better overall throughput.
+	src := stream.Buffer(
+		ListerWatcherToObservable(r.ListerWatcher),
+		bufferSize,
+		waitTime,
+
+		// Buffer the events into a map, coalescing them by key.
+		func(buf *buffer, ev CacheStoreEvent) *buffer {
+			switch {
+			case ev.Kind == CacheStoreEventReplace:
+				return &buffer{
+					replaceItems: ev.Obj.([]any),
+					entries:      make(map[string]entry, bufferSize), // Forget prior entries
+				}
+			case buf == nil:
+				buf = &buffer{
+					replaceItems: nil,
+					entries:      make(map[string]entry, bufferSize),
+				}
+			}
+
+			var entry entry
+			entry.deleted = ev.Kind == CacheStoreEventDelete
+
+			var key string
+			if d, ok := ev.Obj.(cache.DeletedFinalStateUnknown); ok {
+				key = d.Key
+				var err error
+				entry.namespace, entry.name, err = cache.SplitMetaNamespaceKey(d.Key)
+				if err != nil {
+					panic(fmt.Sprintf("%T internal error: cache.SplitMetaNamespaceKey(%q) failed: %s", r, d.Key, err))
+				}
+				entry.obj, ok = transform(d.Obj)
+				if !ok {
+					return buf
+				}
+			} else {
+				meta, err := meta.Accessor(ev.Obj)
+				if err != nil {
+					panic(fmt.Sprintf("%T internal error: meta.Accessor failed: %s", r, err))
+				}
+				entry.name = meta.GetName()
+				if ns := meta.GetNamespace(); ns != "" {
+					key = ns + "/" + meta.GetName()
+					entry.namespace = ns
+				} else {
+					key = meta.GetName()
+				}
+
+				var ok bool
```

Comment:
This looks very similar to lines `175-178`. Is it possible to deduplicate it by moving outside the if-else?

Reply:
Oh wait a minute. `DeleteFinalStateUnknown` is emitted by DeltaFIFO, but since I'm using the client-go reflector directly here I wouldn't ever see it here. I'll remove that branch completely.